streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -552,28 +554,27 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
         return doJoin(other, joiner, windows, joined, new KStreamImplJoin(true, true));
     }
 
-    private <V1, R> KStream<K, R> doJoin(final KStream<K, V1> other,
-                                         final ValueJoiner<? super V, ? super V1, ? extends R> joiner,
-                                         final JoinWindows windows,
-                                         final Joined<K, V, V1> joined,
-                                         final KStreamImplJoin join) {
+    private <VO, VR> KStream<K, VR> doJoin(final KStream<K, VO> other,
+                                           final ValueJoiner<? super V, ? super VO, ? extends VR> joiner,
+                                           final JoinWindows windows,
+                                           final Joined<K, V, VO> joined,
+                                           final KStreamImplJoin join) {
         Objects.requireNonNull(other, "other KStream can't be null");
         Objects.requireNonNull(joiner, "joiner can't be null");
         Objects.requireNonNull(windows, "windows can't be null");
         Objects.requireNonNull(joined, "joined can't be null");
 
         KStreamImpl<K, V> joinThis = this;
-        KStreamImpl<K, V1> joinOther = (KStreamImpl<K, V1>) other;
+        KStreamImpl<K, VO> joinOther = (KStreamImpl<K, VO>) other;
 
         if (joinThis.repartitionRequired) {
             final String leftJoinRepartitionTopicName = joined.name() != null ? joined.name() + "-left" : joinThis.name;
-            joinThis = joinThis.repartitionForJoin(Joined.with(joined.keySerde(), joined.valueSerde(), joined.otherValueSerde(), leftJoinRepartitionTopicName));
+            joinThis = joinThis.repartitionForJoin(leftJoinRepartitionTopicName, joined.keySerde(), joined.valueSerde());
         }
 
         if (joinOther.repartitionRequired) {
             final String rightJoinRepartitionTopicName = joined.name() != null ? joined.name() + "-right" : joinOther.name;
-            final Joined newJoined = Joined.with(joined.keySerde(), joined.valueSerde(), joined.otherValueSerde(), rightJoinRepartitionTopicName);
-            joinOther = joinOther.repartitionForJoin(newJoined);
```

Comment:
I think this was a bug. `repartitionForJoin` actually ignored the `otherValueSerde` and only used the `valueSerde`. In this case, we would use the incorrect serde to de/serialize the repartition topic.

I've clarified it by adding the correct generic type parameters and expanding the argument list of `repartitionForJoin` to only take the information it will actually use.

Reply:
Good catch, but I still prefer to use `Joined` as a single parameter, but overall I think this approach is better.

In the Serde inheritance PR the order was switched when creating a new `Joined` object ie `Joined.with(keySerde, otherValueSerde, valueSerde)` so the `otherValueSerde` was used,  but the change was subtle and I missed that point.  Probably better to keep it this way so things are more explicit.

Reply:
I agree, actually before @bbejeck 's PR the line was 

```
joinOther.repartitionForJoin(Joined.with(joined.keySerde(), joined.otherValueSerde(), joined.valueSerde()));
```

to intentionally switch other value serde and value serde, but the pervious PR dropped this.

Reply:
Thanks, all. I'm sorry, but I'm confused about what you're both suggesting to do.
It sounds like this was a regression during refactoring, which makes sense (and makes me feel better that this bug wasn't released).

But are you suggesting to go back to just the `Joined` parameter for `repartitionForJoin`? Or are you ok with this proposal to have two params "repartition name" and "repartition serde"?

Personally, I'm weakly in favor of the current implementation in my PR, since:
* it doesn't require the construction of a new Joined that will just get discarded
* in particular, Joined passes 3 pieces of information, but we only need 2 of them,
* it also allows us to use the type system to prevent another regression of this nature in the future

Reply:
My apologies for not being clear before. 

I agree with you that I'm in favor of the approach you have here as it's more clear what is going on.  
My original point was I favor the conciseness of a single `Joined` parameter, but it's easier to create a regression (which my PR did), so I vote to pass three parameters to `repartitionForJoin`.

Reply:
Ah, yes, then I agree on both points.

Reply:
> I agree, actually before @bbejeck 's PR the line was

Just clarify the `I agree` was addressing the `I think it is a bug` comment (I think @bbejeck and I were commenting at the same time so these our first two comments were actually in parallel :P). With `Joined` we basically need to do what originally in this code, i.e. creating a newJoined object with the serdes switched. I'm fine with just passing in parameters directly but I'll leave that to your call.